experimental/inductor.py

```diff
@@ -0,0 +1,669 @@
+        return tensor._placements
+    elif isinstance(tensor, DTensor):
+        return tensor.placements
+    else:
+        return None
+
+
+def get_example_shard(tensor, mesh, default=None, fn=torch.randn):
+    if not isinstance(tensor, Tensor):
+        return tensor
+    elif isinstance(tensor, DTensor):
+        return tensor.to_local()
+    elif hasattr(tensor, '_placements'):
+        inp_spec = DTensorSpec(mesh, tensor._placements, tensor.shape, tensor.ndim)
+        return fn(inp_spec.local_shape, device=tensor.device)
+    elif default is not None:
+        inp_spec = DTensorSpec(mesh, default, tensor.shape, tensor.ndim)
+        return fn(inp_spec.local_shape, device=tensor.device)
+    raise NotImplementedError(f'Dont know how to get a shard of tensor {tensor}')
+
+
+def convert_getattrs_to_inputs(gm):
+    """
+    Turn getattr into inputs
+        - This is needed for functionalize to work. Functionalize expects
+          all tensors to be passed as inputs to the function being functionalized.
+        - We collect all concrete getattr inputs into "attr_vals"
+    """
+    from collections import defaultdict
+    attr_vals_map = dict()
+    attr_nodes = defaultdict(lambda: [])
+
+    for p in gm.graph.nodes:
+        if p.op == 'get_attr':
+            val = getattr(gm, p.target)
+            attr_vals_map[p.target] = val
+            attr_nodes[p.target].append(p)
+
+    after_last_placeholder = None
+    for i, p in enumerate(gm.graph.nodes):
+        if p.op != 'placeholder':
+            after_last_placeholder = p
+            break
+
+    attr_vals = []
+    with gm.graph.inserting_before(after_last_placeholder):
+        for attr, nodes in attr_nodes.items():
+            placeholder = gm.graph.placeholder(attr)
+            attr_vals.append(attr_vals_map[attr])
+            for node in nodes:
+                node.replace_all_uses_with(placeholder)
+                gm.graph.erase_node(node)
+
+    return attr_vals
+
+
+def make_functional_fx(fn):
+    """
+    Given a function, does make_fx(functionalize(make_fx(fn))).
+    Some input-related transformations need to be done before functionalize.
+    Returns an GraphModule to be fed to inductor.
+    """
+    def call(*args, **kwargs):
+        # 0. hack: run one iteration to initialize optimizer states
+        for inp in args:
+            if isinstance(inp, torch.optim.Optimizer):
+                warmup_optimizer(inp)
+
+        # 1. call make_fx for the first time
+        k1 = make_fx(fn)
+        k2 = k1(*args, **kwargs)
+        print_graph('first_fx', k2)
+
+        attr_vals = convert_getattrs_to_inputs(k2)
+        print_value('attr_vals', attr_vals)
+
+        # 3. Remove unused inputs
+        #    - This is needed in order to remove non-tensor inputs before passing this
+        #      graph to inductor. Inductor expects only tensors as inputs
+        #    - This works since all the non-tensor inputs should be unused at this point,
```

Comment:
will there be any problems for non-tensor inputs? i.e. forward that contains both a tensor input and some non-tensor args

Reply:
Yes this isn't battle proof yet. Right now I didn't flex the types of things you pass in.
Also, in general, since we're bypassing dynamo, there's no guarantees that non-tensor ops will be properly captured, as they would ended up being baked into the graph.
Ultimately it would be best if dynamo could produce a single graph without breaks. Right now it's breaking on .backwards()